circuits/cpp/src/aztec3/circuits/kernel/public/common.hpp

```diff
@@ -0,0 +1,178 @@
+                         public_kernel_inputs.public_call.public_data_tree_root,
+                         msg);
+    }
+};
+
+template <typename KernelInput>
+void validate_this_public_call_stack(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Ensures that the stack of pre-images corresponds to the call stack
+    auto& stack = public_kernel_inputs.public_call.public_call_data.call_stack_item.public_inputs.public_call_stack;
+    auto& preimages = public_kernel_inputs.public_call.public_call_data.public_call_stack_preimages;
+    for (size_t i = 0; i < stack.size(); ++i) {
+        const auto& hash = stack[i];
+        const auto& preimage = preimages[i];
+
+        // Note: this assumes it's computationally infeasible to have `0` as a valid call_stack_item_hash.
+        // Assumes `hash == 0` means "this stack item is empty".
+        const auto calculated_hash = hash == 0 ? 0 : preimage.hash();
+        composer.do_assert(hash != calculated_hash,
+                           format("public_call_stack[", i, "] = ", hash, "; does not reconcile"));
+    }
+};
+
+template <typename KernelInput>
+void validate_function_execution(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Validates state reads and transitions for all type of kernel inputs
+    validate_state_reads(composer, public_kernel_inputs);
+    validate_state_transitions(composer, public_kernel_inputs);
+}
+
+template <typename KernelInput>
+void common_validate_kernel_execution(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Validates kernel execution for all type of kernel inputs
+    validate_this_public_call_stack(composer, public_kernel_inputs);
+
+    validate_function_execution(composer, public_kernel_inputs);
+}
+
+template <typename KernelInput>
+void common_validate_inputs(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Validates commons inputs for all type of kernel inputs
+    const auto& this_call_stack_item = public_kernel_inputs.public_call.public_call_data.call_stack_item;
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.is_contract_deployment == false,
+                       "Contract deployment can't be a public function");
+    composer.do_assert(this_call_stack_item.contract_address != 0, "Contract address must be valid");
+    composer.do_assert(this_call_stack_item.function_data.function_selector != 0, "Function signature must be valid");
+    composer.do_assert(this_call_stack_item.function_data.is_constructor == false,
+                       "Constructors can't be public functions");
+    composer.do_assert(this_call_stack_item.function_data.is_private == false,
+                       "Cannot execute a private function with the public kernel circuit");
+    composer.do_assert(public_kernel_inputs.public_call.public_call_data.bytecode_hash != 0,
+                       "Bytecode hash must be valid");
```

Comment:
Nice, clean. I like how you've organized these public kernel functions into these common helpers.